### DISCUSSION TITLE
Placing multiple  widgets

### DISCUSSION THREAD

**Discussion by vikash0705** [mum4k/termdash#320] [General]
Can we place multiple widgets inside a container vertically or horizontally without splitting it?

**Comment by mum4k** [comment]
Hi @vikash0705, the Termdash infrastructure only supports a single widget in a container. The splitting methodology exists to allow the placement of multiple widgets while keeping the size of individual layout elements relative to the size of the terminal window.

Note that there are two layouts that split the terminal into containers:

- [Binary tree layout](https://github.com/mum4k/termdash/wiki/Binary-tree-layout)
- [Grid layout](https://github.com/mum4k/termdash/wiki/Grid-layout)

Maybe one of these two approaches addresses the need? Alternatively, if you could provide more details about what you are trying to achieve, I can try to outline if and how this can be done using one of the above layouts.

**Reply by vikash0705** [reply]
Basically I have a list of 13(say) text widgets which i want to place horizontally one after the other, on the left container if say parent container is divided vertically. So how can we achieve that?

**Comment by mum4k** [comment]
A layout like that should be possible. I would recommend using the grid builder. The layout is in essence similar to the one created by this [unit test](https://github.com/mum4k/termdash/blob/f13e4222fa62b251e937fa2a00ffab003d814f13/container/grid/grid_test.go#L823-L854) (i.e. these are columns with inner rows).

Can you give it a try and indicate if you hit any issues?

**Reply by vikash0705** [reply]
Sure will try, also I am facing one issue with widgets getting hidden. This is happening if my content overflows the size the of the terminal. Do we have a scrollable layout or how to enable it in grid or tree layout?

**Reply by mum4k** [reply]
The infrastructure currently doesn't support scrolling (https://github.com/mum4k/termdash/issues/192). The widget gets hidden when the available space is lower than what the widget needs to display some meaningful content. This is [configurable](https://github.com/mum4k/termdash/blob/f13e4222fa62b251e937fa2a00ffab003d814f13/widgetapi/widgetapi.go#L115-L126) in each widget.

Have you considered a form of [dynamic layout](https://github.com/mum4k/termdash/wiki/Dynamic-layout) - where not all widgets will be displayed at the same time? Clicking on elements or buttons could change the list of displayed widgets, similar to what the [main demo](https://github.com/mum4k/termdash/blob/master/README.md) does.

**Comment by vikash0705** [comment]
So I have a problem statement like , will have and Rest API , which would be serving some list of json objects now the length can be varying , so I am thinking of having a text box here I enter a unique reference of a given json object and then its data would be shown in a text widget. Any suggestions how to achieve this?

**Reply by mum4k** [reply]
Thank you for explaining that @vikash0705. I would consider a fixed terminal layout even though the number of the json objects is variable. This is because if we would get too many json objects and the terminal layout is dynamic (displaying all of them at once), the individual text widgets could end up being too small.

We could choose a fixed layout with a number of text widgets based on preference. Then we could add buttons or keyboard shortcuts that would "scroll" the displayed json objects by one or by a page. Each time the button or the key is pressed, the context of the text widgets shifts - some json objects would be displayed and some others would hide. There could also be another static text widget displaying the full size and the current position.

In essence this would be an implementation of "paging". Not sure if this advice helps, please let me know if you have another question.

**Reply by vikash0705** [reply]
So I have divided the terminal into 4 boxes each having a text widgets, now i have to keep on calling an api in the background at every 5 secs, and then have to distribute the response and write it on the text widgets every time. Do you have an approach for this?

**Reply by mum4k** [reply]
That sounds great @vikash0705, I am glad you found a layout that works.

Can you help me understand the question? What you are describing does sound like a workable solution. There can be a goroutine that periodically fetches the information and updates the context of the text widgets. Did you encounter some problem while setting that up?
